core/modules/replica/DatabaseServicesPool.cc

```diff
@@ -0,0 +1,322 @@
+    // in order to release allocated services.
+
+    auto service = allocateService();
+    try {
+        service->findReplicas(replicas,
+                              chunk,
+                              database,
+                              enabledWorkersOnly);
+        releaseService(service);
+
+    } catch (std::exception const& ex) {
+        releaseService(service);
+        throw;
+    }
+}
+
+void DatabaseServicesPool::findWorkerReplicas(std::vector<ReplicaInfo>& replicas,
+                                              std::string const& worker,
+                                              std::string const& database) {
+
+    // IMPLEMENTATION NOTE: exceptions are temporary intercepted (and re-thrown)
+    // in order to release allocated services.
+
+    auto service = allocateService();
+    try {
+        service->findWorkerReplicas(replicas,
+                                    worker,
+                                    database);
+        releaseService(service);
+
+    } catch (std::exception const& ex) {
+        releaseService(service);
+        throw;
+    }
+}
+
+void DatabaseServicesPool::findWorkerReplicas(std::vector<ReplicaInfo>& replicas,
+                                              unsigned int chunk,
+                                              std::string const& worker,
+                                              std::string const& databaseFamily) {
+
+    // IMPLEMENTATION NOTE: exceptions are temporary intercepted (and re-thrown)
+    // in order to release allocated services.
+
+    auto service = allocateService();
+    try {
+        service->findWorkerReplicas(replicas,
+                                    chunk,
+                                    worker,
+                                    databaseFamily);
+        releaseService(service);
+
+    } catch (std::exception const& ex) {
+        releaseService(service);
+        throw;
+    }
+}
+
+DatabaseServices::Ptr DatabaseServicesPool::allocateService() {
+
+    std::string const context = "DatabaseServicesPool::allocateService  ";
+
+    LOGS(_log, LOG_LVL_DEBUG, context);
+
+    std::unique_lock<std::mutex> lock(_mtx);
+
+    auto self = shared_from_base<DatabaseServicesPool>();
+
+    _available.wait(lock, [self]() {
+        return not self->_availableServices.empty();
+    });
+    
+    // Get the next request and move it between queues.
+
+    DatabaseServices::Ptr service = _availableServices.front();
+    _availableServices.pop_front();
+    _usedServices.push_back(service);
+
+    return service;
+}
+
+void DatabaseServicesPool::releaseService(DatabaseServices::Ptr const& service) {
+
+    std::string const context = "DatabaseServicesPool::releaseService  ";
+
+    LOGS(_log, LOG_LVL_DEBUG, context);
+
+    std::unique_lock<std::mutex> lock(_mtx);
+
+    // Move it between queues.
+
+    _usedServices.remove(service);
+    _availableServices.push_back(service);
```

Comment:
Reinforced to check if the specified point is found in the collection, and throw an exception if not. This is just in case. I shouldn't happen with the current implementation. Still it's good to have this check in place if someone modifies a client code and attempts to double-release a service.